### PR TITLE
Use JSTL 2.0 TLDs in Pages 3.0

### LINKED ISSUE #15317: Use EE9 JSTL TLDs for Pages-3.0
Currently, the com.ibm.ws.jsp bundle uses the JSTL TLDs from version 1.2.  This means all 3 JSP features( jsp-2.2, jsp-2.3, and pages-3.0) use the same files. Pages-3.0 does require a few transformations: https://github.com/OpenLiberty/open-liberty/blob/625d50b068734b9bd52c1398b20374eac6b6c9db/dev/wlp-jakartaee-transform/rules/jakarta-tld.properties#L1-L5

This is still the case even with the JSTL Refactoring PR #15141. 

These current TLD files are pulled in from 
https://github.com/OpenLiberty/open-liberty/blob/9974f20f4c8930c981071b7d01dd1581b537db14/dev/com.ibm.ws.jsp/bnd.bnd#L124
https://github.com/OpenLiberty/open-liberty/blob/c68c2dd61f25397a85fc9c1bb7aea85c8af084a4/dev/cnf/oss_ibm.maven#L23
(note the standardtld.jar is renamed to standard.jar as it's placed within the com.ibm.ws.jsp bundle) 

_However, if future versions of JSTL 2.0 have new additions to these TLD files which the transformer cannot handle, then this will pose a problem for our Pages-3.0 feature._

One potential solution:  

1.  Create a EE9 standardtld.jar. 
2.  Include in within the com.ibm.ws.jsp bundle (which would then contain both versions of the tld jars) 
3.  Have the TLD processing read the appropriate jar depending which JSP feature is enabled (by checking the loaded JSP version).

Relevant portions of code related to TLD Processing include:
https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/GlobalTagLibraryCache.java#L130
https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/config/TagLibCacheConfigParser.java#L33

### PR DESCRIPTION
**Pull request by volosied**
fixes #15317

This PR
- Creates **io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal** jar which only contains the new Jakarta TLDs.
- Includes a new jar in _com.ibm.ws.jsp_ named as  _standard.tags.2.0.jar_   (contains the TLDs from the bundle above) 
- Removes previous transformations as those aren't necessary anymore. 

During JSP initalization, the new code loads the proper TLDs for the EE8 & EE9 features by checking the JSPExtensionFactory#loadedSpecLevel. 

Only hiccup with this change was JSPExtensionFactory.setVersion & JSPExtensionFactory.setGlobalTagLibConfig

When setGlobalTagLibConfig is called, it relies on the current JSP Version. However, the version is not yet set since setVersion is called after setGlobalTagLibConfig.  We cache the globalTagLibConfig argument and call setGlobalTagLibConfig later on from within the setVersion method.  A similar comment was added in the code.

### CODE REVIEW

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
Build failed due to...  Not sure why. Maybe it didn't have the JSP Version change pulled in? Locally, jsp/pages seems to work fine. Running another build.   
```
	FAILED
    [junit] Found issues in console.log after start: [[WARNING ] jsp.feature.not.loaded.correctly]
    [junit] junit.framework.AssertionFailedError: Found issues in console.log after start: [[WARNING ] jsp.feature.not.loaded.correctly]
```

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/GlobalTagLibraryCache.java [hunk 1]
```diff
@@ -127,7 +127,22 @@ public GlobalTagLibraryCache(ClassLoader loader) {
             tldParser = new TldParser(this, configManager, false);
             
             TagLibCacheConfigParser tagLibCacheConfigParser = new TagLibCacheConfigParser();
-            tagLibCacheConfigParser.parse(this.getClass().getResourceAsStream("/com/ibm/ws/jsp/resources/taglibcacheconfig.xml"));
+
+            String pagesVersion = com.ibm.ws.jsp.webcontainerext.JSPExtensionFactory.getLoadedPagesSpecLevel();
+            switch(pagesVersion){
+                case "3.0":
+                    if(com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled()&&logger.isLoggable(Level.FINE)){
+                        logger.logp(Level.FINE, CLASS_NAME, "GlobalTagLibraryCache", "Loading taglibcacheconfig.jstl.2.0.xml for pages version " + pagesVersion);
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/GlobalTagLibraryCache.java:135]: Maybe change wording. Parsing over loading.

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java [hunk 2]
```diff
@@ -196,6 +201,11 @@ protected void modified(Map<Object, Object> properties) {
     protected synchronized void setVersion(ServiceReference<PagesVersion> reference) {
       this.versionRef = reference;
       JSPExtensionFactory.loadedSpecLevel = (String) reference.getProperty("version");
+      isPagesVersionLoaded = true;
+      if(!storedGlobalTagLibConfigs.isEmpty()){
+        storedGlobalTagLibConfigs.forEach(tagConfig ->  getGlobalTagLibraryCache().addGlobalTagLibConfig(tagConfig));
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java]: I add the addGlobalTagLibConfig here instead. (see the other comments below). 

I could just call setGlobalTagLibConfig instead?

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java [hunk 3]
```diff
@@ -116,6 +117,10 @@
     private static String loadedSpecLevel = SPEC_LEVEL_UNLOADED;
 
     protected static volatile CountDownLatch selfInit = new CountDownLatch(1);
+
+    // Add OL Issue # Here If This Approach is Accepted
+    private boolean isPagesVersionLoaded = false;
+    ArrayList<GlobalTagLibConfig> storedGlobalTagLibConfigs = new ArrayList<GlobalTagLibConfig>();
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java:123]: I believe there's only one used, but I created an ArrayList in case.

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.jsp/bnd.bnd [hunk 4]
```diff
@@ -157,4 +159,5 @@ instrument.classesExcludes: com/ibm/ws/jsp/resources/*.class
 	com.ibm.ws.artifact.url;version=latest, \
 	com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
 	com.ibm.ws.org.apache.taglibs.standard;version=latest, \
-	com.ibm.ws.javaee.version;version=latest
+	com.ibm.ws.javaee.version;version=latest, \
+   io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal;version=latest
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/bnd.bnd:163]: Fix indent

Comment (pnicolucci) [reply dev/com.ibm.ws.jsp/bnd.bnd:163]: @volosied  please do fix the formatting

File: dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/taglibcacheconfig.jstl.2.0.xml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2017 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/taglibcacheconfig.jstl.2.0.xml:4]: This looks like a new file, why do we have a 2017 copyright date?

File: dev/com.ibm.ws.jsp/bnd.bnd [hunk 6]
```diff
@@ -112,10 +112,12 @@ Import-Package: \
 Include-Resource: \
    @lib/bsf.jar, \
    standard.jar=${repo;com.ibm.ws.javax.j2ee:standardtld;1.2}, \
+   standard.jstl.2.0.jar=${repo;io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal}, \
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/bnd.bnd]: `standard.jstl.2.0.jar` is the jar within the bundle that contains the new TLDs

File: dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/taglibcacheconfig.jstl.2.0.xml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		<prefix>tsx</prefix>
+		<location>/com/ibm/ws/jsp/tsx/tag/tsx.tld</location>
+	</implict-taglib>
+	<implict-taglib>
+		<uri>http://websphere.ibm.com/tags/jsx</uri>
+		<prefix>jsxFunction</prefix>
+		<location>/com/ibm/ws/jsp/jsx/tags/jsxFunction.tld</location>
+	</implict-taglib>
+	<global-taglib>
+		<jar-name>standard.jstl.2.0.jar</jar-name>
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/taglibcacheconfig.jstl.2.0.xml]: Same as the original taglibcacheconfig.xml ([link](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/taglibcacheconfig.xml))  but contains a reference to the new jar here on line 26.

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/GlobalTagLibraryCache.java [hunk 8]
```diff
@@ -127,7 +127,19 @@ public GlobalTagLibraryCache(ClassLoader loader) {
...
+                    logger.logp(Level.FINE, CLASS_NAME, "GlobalTagLibraryCache", "Parsing taglibcacheconfig.jstl.2.0.xml for pages version >= 3.0");
+                }
+                tagLibCacheConfigParser.parse(this.getClass().getResourceAsStream("/com/ibm/ws/jsp/resources/taglibcacheconfig.jstl.2.0.xml"));
+            } else {
+                if(com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled()&&logger.isLoggable(Level.FINE)){
+                    logger.logp(Level.FINE, CLASS_NAME, "GlobalTagLibraryCache", "Parsing taglibcacheconfig.xml for pages version <= 2.3");
+                }
+                tagLibCacheConfigParser.parse(this.getClass().getResourceAsStream("/com/ibm/ws/jsp/resources/taglibcacheconfig.xml"));
+            }
+
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/GlobalTagLibraryCache.java]: Parses the correct taglibcacheconfig file based on which pages version is loaded.

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/config/TagLibCacheConfigParser.java [hunk 9]
```diff
@@ -171,7 +174,12 @@ public InputSource resolveEntity(String publicId, String systemId)
         InputSource isrc = null;
         String resourcePath = null;            
         if (publicId.equals(DTD_PUBLIC_ID)) {
-            resourcePath = DTD_RESOURCE_PATH;
+
+            if(com.ibm.ws.jsp.webcontainerext.JSPExtensionFactory.isPages30orHigher()){
+                resourcePath = DTD_RESOURCE_PATH_EE9;
+            } else {
+                resourcePath = DTD_RESOURCE_PATH;
+            }
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/config/TagLibCacheConfigParser.java]: Once again, making sure the right taglibcacheconfig is used during parsing.

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java [hunk 10]
```diff
@@ -598,7 +608,12 @@ public static synchronized GlobalTagLibraryCache getGlobalTagLibraryCache() {
      */
     @Reference(cardinality=ReferenceCardinality.MULTIPLE, policy=ReferencePolicy.DYNAMIC)
     protected void setGlobalTagLibConfig(GlobalTagLibConfig globalTagLibConfig) {
-            getGlobalTagLibraryCache().addGlobalTagLibConfig(globalTagLibConfig);
+            // Wait for pages version to be loaded first before calling getGlobalTagLibraryCache
+            if(!isPagesVersionLoaded){
+                storedGlobalTagLibConfigs.add(globalTagLibConfig);
+            } else {
+                getGlobalTagLibraryCache().addGlobalTagLibConfig(globalTagLibConfig);
+            }
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java]: When `addGlobalTagLibConfig` is called,  it eventually relies on the loaded pages version. However, it's not yet set  since `setGlobalTagLibConfig` is called before `setVersion`.   We cache the `globalTagLibConfig` argument and then call addGlobalTagLibConfig later in the `setVersion` call. 

I spoke to Tom Watson, and  we aren't aware of a way to change the order here.

File: dev/io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal/bnd.bnd [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+bVersion=1.0
+
+Bundle-SymbolicName: io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal
+Bundle-Description: TLDs for the Jakarta Standard Tag Library; Version 2.0.0
+
+-includeresource: \
+	@${repo;org.glassfish.web:jakarta.servlet.jsp.jstl;2.0.0}!/META-INF/*.tld
+
+-buildpath: \
+  org.glassfish.web:jakarta.servlet.jsp.jstl;version=2.0
```
Comment (volosied) [inline dev/io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal/bnd.bnd:21]: New bundle which only contains the TLD resources.  These TLDs are later added in the standard.jstl.2.0.jar.

Hmm. I might be able to avoid this bundle and directly export the TLDs  from the API into standard.jstl.2.0.jar. I'll test it locally.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

File: dev/io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal/bnd.bnd (hunk 11 above)
Comment (pnicolucci) [reply dev/io.openliberty.jakarta.servlet.jsp.jstl.2.0.internal/bnd.bnd:21]: Were you ever able to get the direct export working?

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

File: dev/wlp-jakartaee-transform/rules/jakarta-xml-master.properties [hunk 12]
```diff
@@ -65,6 +62,7 @@ com.ibm.ws.jca.processor.jms.service.JMSConnectionFactoryResourceBuilder.xml=jak
 com.ibm.ws.jca.processor.jms.service.JMSDestinationResourceFactoryBuilder.xml=jakarta-resource.properties
 # Pages
 com.ibm.ws.jsp.v23.dd.xml=jakarta-jspVersion.properties
+SerializationTestJSP.jsp=jakarta-renames.properties
 #Faces
```
Comment (pnicolucci) [inline dev/wlp-jakartaee-transform/rules/jakarta-xml-master.properties]: Can you explain why this change is needed for this particular set of changes?

**Comment by volosied** [comment]
#build

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
#run-libby-bot
